feat(adapters): add Hasura adapter

arjunyel opened the pull request:
## ☕️ Reasoning

Adding a Hasura adapter. The Hasura team is willing to maintain

## 🧢 Checklist

- [X] Documentation
- [X] Tests
- [X] Ready to be merged

arjunyel committed:
feat(adapters): add Hasura adapter

arjunyel committed:
merge main

arjunyel committed:
Merge branch 'main' into hasura-adapter

arjunyel committed:
Merge branch 'main' into hasura-adapter

arjunyel commented:
Hi friends, anything you need from me to land this? A lot of the code is auto-generated and tree shaken when you build :)

balazsorban44 commented:
Hi, just wanted to give a note, this is interesting for us, just haven't had the time to look into it yet! :bow:

> "The Hasura team is willing to maintain"

:green_heart:

azlekov commented:
What is the plan here? Any timeframe when the Hasura adapter will be official?

ThangHuuVu committed:
Merge branch 'main' into hasura-adapter

ThangHuuVu committed:
chore: formatting

ThangHuuVu committed:
chore: formatting

ThangHuuVu committed:
chore: formatting

ThangHuuVu committed:
chore: formatting

ThangHuuVu commented:
hi @arjunyel , I pushed some commits to format the PR. 

- [x] Publish it as ESM-only and under @auth/ namespace. Make sure to have correct package.json file and tsconfig.json file
- [x] Add documentation as JSDoc.

To get this PR merged, could you help with the build errors? Sorry, I am still not very familiar yet with Hasura & GraphQL so I didn't touch the code in `src/`.
 I'm curious about the files that you added in the PR. Could you walk me through all the files? I am guessing that we don't need the `gql` and `hasura` directories, but not sure 🤔. Thanks for the PR & sorry for the delay, let's get this done!

arjunyel committed:
Merge branch 'main' into hasura-adapter

arjunyel committed:
feat(adapter): add Hasura adapter

arjunyel committed:
Merge branch 'main' into hasura-adapter

arjunyel commented:
@ThangHuuVu Fixed build error and cleaned up.

We need the `hasura` directory in order to run the tests against a Hasura instance. `gql` directory has been renamed to `lib` and is the output of GraphQL Codegen, basically it's a generated SDK for accessing Hasura.

Btw I am no longer on the Hasura team so I am not sure of their support plans of this. Thanks for all your help friend!

arjunyel committed:
chore: update Hasura adapter readme

arjunyel committed:
chore(docs): add Hasura

ThangHuuVu commented:
> Btw I am no longer on the Hasura team so I am not sure of their support plans of this. Thanks for all your help friend!

🫶 

> We need the `hasura` directory in order to run the tests against a Hasura instance. `gql` directory has been renamed to `lib` and is the output of GraphQL Codegen, basically it's a generated SDK for accessing Hasura.

Thanks for the explanation - if we need to ship the codegen output as part of the library, let's run the script before running `tsc`?

ThangHuuVu commented on packages/adapter-hasura/src/index.ts:
```diff
@@ -0,0 +1,498 @@
+} from "./lib/graphql"
+import { formatDateConversion } from "./utils"
+import type { NonNullify } from "./utils"
+import { log } from "console"
```
unused import, let's remove

ThangHuuVu commented on packages/adapter-hasura/src/index.ts:
````diff
@@ -0,0 +1,498 @@
+ *       ADD CONSTRAINT "accounts_type_fkey" FOREIGN KEY ("type") REFERENCES public.provider_type(value) ON UPDATE RESTRICT ON DELETE RESTRICT;
+ *   ```
+ *
+ *1. [Track all the tables and relationships in Hasura](https://hasura.io/docs/latest/schema/postgres/using-existing-database/#step-1-track-tablesviews)
````
Let's write this as a Info block:
:::info
Tips: [Track all the tables and relationships in Hasura](https://hasura.io/docs/latest/schema/postgres/using-existing-database/#step-1-track-tablesviews)
:::

ThangHuuVu commented on packages/adapter-hasura/src/index.ts:
````diff
@@ -0,0 +1,498 @@
+ *
+ *1. [Track all the tables and relationships in Hasura](https://hasura.io/docs/latest/schema/postgres/using-existing-database/#step-1-track-tablesviews)
+ *
+ *1. Install the necessary packages
+ *
+ *   ```bash npm2yarn2pnpm
+ *   npm install next-auth @next-auth/hasura-adapter graphql graphql-request
+ *   ```
````
Let's remove this part, we already have a installation section above

arjunyel committed:
Merge remote-tracking branch 'origin/main' into hasura-adapter

arjunyel committed:
Merge branch 'main' into hasura-adapter

arjunyel committed:
feat(adapter): move Hasura codegen to script

arjunyel commented:
@ThangHuuVu addressed feedback and moved codegen to a build script

arjunyel committed:
Merge branch 'main' into hasura-adapter

ThangHuuVu commented:
@arjunyel could you help check the build error, I wonder if we really need docker for code gen? 🤔

arjunyel committed:
feat(adapter): remove docker from Hasura build

arjunyel commented:
@ThangHuuVu sorry for all the trouble! Removed Docker from the build step, replaced it by having the GraphQL Schema as a file in the repo. I need the schema for the codegen so its either run a docker container or have the schema in a GraphQL file. Thank you for all the help!

ThangHuuVu committed:
Merge branch 'main' into hasura-adapter

ThangHuuVu committed:
chore: resolve conflict

ThangHuuVu committed:
Merge branch 'main' into hasura-adapter

ThangHuuVu committed:
fix test

ThangHuuVu committed:
fix test
